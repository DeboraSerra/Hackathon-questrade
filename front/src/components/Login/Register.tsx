"use client";
import {
  parseCpf,
  parsePhone,
  validateCpf,
  validatePassword,
} from "@/lib/helpers";
import { useRouter } from "next/navigation";
import {
  ChangeEventHandler,
  FormEventHandler,
  Suspense,
  useState,
} from "react";
import { BsEye, BsEyeSlash } from "react-icons/bs";
import Spinner from "../CommonLayout/Spinner";

const Register = ({ setRegister }: { setRegister: (val: boolean) => void }) => {
  const [show, setShow] = useState(false);
  const [form, setForm] = useState({
    name: "",
    email: "",
    password: "",
    confirmPassword: "",
    cpf: "",
    phone: "",
    proofOfIncome: "",
  });
  const [hasError, setHasError] = useState<boolean | string>(false);

  const router = useRouter();

  const handleSubmit: FormEventHandler = (e) => {
    e.preventDefault();
    router.push("/dashboard");
  };

  const handleChange: ChangeEventHandler = (e) => {
    const { name } = e.target as HTMLInputElement;
    let { value } = e.target as HTMLInputElement;
    if (name === "phone") {
      value = parsePhone(value);
    }
    if (name === "cpf") {
      const isValid = validateCpf(value);
      value = parseCpf(value);
      isValid ? setHasError(false) : setHasError("cpf");
      console.log({ name, value, isValid, hasError });
    }
    if (name === "password") {
      const isValid = validatePassword(value);
      !isValid ? setHasError("password") : setHasError(false);
    }
    if (name === "confirmPassword") {
      value !== form.password && setHasError("confirmPassword");
    }
    setForm((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  return (
    <Suspense fallback={<Spinner />}>
      <form
        onSubmit={handleSubmit}
        className="my-3 flex w-[400px] flex-col justify-between gap-4 border px-3 py-4 max-sm:mx-3 max-sm:w-full"
      >
        <h2 className="text-lg font-bold">Cadastro</h2>
        <label htmlFor="name" className="flex flex-col gap-2 text-sm">
          Nome
          <input
            type="text"
            name="name"
            id="name"
            value={form.name}
            onChange={handleChange}
            className="rounded-md border-none bg-gray-50"
          />
        </label>
        <label htmlFor="email" className="flex flex-col gap-2 text-sm">
          E-mail
          <input
            type="email"
            name="email"
            id="email"
            value={form.email}
            onChange={handleChange}
            className="rounded-md border-none bg-gray-50"
          />
        </label>
        <label htmlFor="cpf" className="relative flex flex-col gap-2 text-sm">
          CPF
          <input
            type="text"
            name="cpf"
            id="cpf"
            value={form.cpf}
            placeholder=" "
            onChange={handleChange}
            className={`rounded-md ${
              hasError === "cpf" ? "border-red-600" : "border-none"
            } bg-gray-50 focus:outline-none`}
          />
          {hasError === "cpf" && (
            <small className="lh-1 absolute -bottom-4">
              CPF precisa ser um número válido
            </small>
          )}
        </label>
        <label htmlFor="phone" className="flex flex-col gap-2 text-sm">
          Telefone
          <input
            type="text"
            name="phone"
            id="phone"
            value={form.phone}
            onChange={handleChange}
            className="rounded-md border-none bg-gray-50"
            maxLength={13}
          />
        </label>
        <label htmlFor="proofOfIncome" className="flex flex-col gap-2 text-sm">
          Comprovação de renda
          <input
            type="text"
            name="proofOfIncome"
            id="proofOfIncome"
            value={form.proofOfIncome}
            onChange={handleChange}
            className="rounded-md border-none bg-gray-50"
          />
        </label>
        <label
          htmlFor="password"
          className="relative flex flex-col gap-2 text-sm"
        >
          <h2 className="text-lg font-bold">Cadastro</h2>
          <label htmlFor="email" className="flex flex-col gap-2 text-sm">
            E-mail
            <input
              type="email"
              name="email"
              id="email"
              value={form.email}
              onChange={handleChange}
              className="rounded-md border-none bg-gray-50"
            />
          </label>
          <label htmlFor="cpf" className="relative flex flex-col gap-2 text-sm">
            CPF
            <input
              type="text"
              name="cpf"
              id="cpf"
              value={form.cpf}
              placeholder=" "
              onChange={handleChange}
              className={`rounded-md ${
                hasError === "cpf" ? "border-red-600" : "border-none"
              } bg-gray-50 focus:outline-none`}
            />
            {hasError === "cpf" && (
              <small className="lh-1 absolute -bottom-4">
                CPF precisa ser um número válido
              </small>
            )}
          </label>
          <label htmlFor="phone" className="flex flex-col gap-2 text-sm">
            Telefone
            <input
              type="text"
              name="phone"
              id="phone"
              value={form.phone}
              onChange={handleChange}
              className="rounded-md border-none bg-gray-50"
              maxLength={13}
            />
          </label>
          <label
            htmlFor="proofOfIncome"
            className="flex flex-col gap-2 text-sm"
          >
            Comprovação de renda
            <input
              type="text"
              name="proofOfIncome"
              id="proofOfIncome"
              value={form.proofOfIncome}
              onChange={handleChange}
              className="rounded-md border-none bg-gray-50"
            />
          </label>
          <label
            htmlFor="password"
            className="relative flex flex-col gap-2 text-sm"
          >
            Senha
            <input
              type={show ? "text" : "password"}
              name="password"
              id="password"
              value={form.password}
              onChange={handleChange}
              className={`rounded-md ${
                hasError === "password" ? "border-red-600" : "border-none"
              } bg-gray-50`}
            />
            <button
              type="button"
              onClick={() => setShow(!show)}
              className="absolute right-2 top-10"
            >
              {show ? <BsEyeSlash /> : <BsEye />}
            </button>
            {hasError === "password" && (
              <small className="lh-1 absolute -bottom-6">
                Senha precisa conter 1 letra minúscula, 1 letra maiúscula, 1
                número e 1 símbolo
              </small>
            )}
          </label>
          <label
            htmlFor="confirmPassword"
            className="relative flex flex-col gap-2 text-sm"
          >
            Confirmar Senha
            <input
              type={show ? "text" : "password"}
              name="confirmPassword"
              id="confirmPassword"
              value={form.confirmPassword}
              onChange={handleChange}
              className={`rounded-md ${
                hasError === "confirmPassword"
                  ? "border-red-600"
                  : "border-none"
              } bg-gray-50`}
            />
            <button
              type="button"
              onClick={() => setShow(!show)}
              className="absolute right-2 top-10"
            >
              {show ? <BsEyeSlash /> : <BsEye />}
            </button>
            {hasError === "confirmPassword" && (
              <small className="lh-1 absolute -bottom-4">
                As senhas precisam ser iguais
              </small>
            )}
          </label>
          <button
            type="submit"
            className="rounded-md bg-green-600 px-6 py-2.5 font-sans text-sm font-semibold text-white"
          >
            CADASTRAR
          </button>
          {hasError === "password" && (
            <small className="lh-1 absolute -bottom-6">
              Senha precisa conter 1 letra minúscula, 1 letra maiúscula, 1
              número e 1 símbolo
            </small>
          )}
        </label>
        <label
          htmlFor="confirmPassword"
          className="relative flex flex-col gap-2 text-sm"
        >
          Confirmar Senha
          <input
            type={show ? "text" : "password"}
            name="confirmPassword"
            id="confirmPassword"
            value={form.confirmPassword}
            onChange={handleChange}
            className={`rounded-md ${
              hasError === "confirmPassword" ? "border-red-600" : "border-none"
            } bg-gray-50`}
          />
          <button
            type="button"
            onClick={() => setShow(!show)}
            className="absolute right-2 top-10"
          >
            {show ? <BsEyeSlash /> : <BsEye />}
          </button>
          {hasError === "confirmPassword" && (
            <small className="lh-1 absolute -bottom-4">
              As senhas precisam ser iguais
            </small>
          )}
        </label>
        <button
          type="submit"
          className="rounded-md bg-green-600 px-6 py-2.5 font-sans text-sm font-semibold text-white"
        >
          CADASTRAR
        </button>
        <p className="text-center font-sans text-xs">
          Já sou cadastrado.{" "}
          <a
            className="cursor-pointer text-blue-100 underline"
            onClick={() => setRegister(false)}
          >
            Fazer login
          </a>
        </p>
      </form>
    </Suspense>
  );
};

export default Register;
