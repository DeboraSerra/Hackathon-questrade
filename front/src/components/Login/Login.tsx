"use client";
import { context } from "@/lib/context";
import { useRouter } from "next/navigation";
import {
  ChangeEventHandler,
  FormEventHandler,
  Suspense,
  useContext,
  useState,
} from "react";
import { BsEye, BsEyeSlash } from "react-icons/bs";
import Spinner from "../CommonLayout/Spinner";
import Register from "./Register";

const Login = () => {
  const [show, setShow] = useState(false);
  const [register, setRegister] = useState(false);
  const [form, setForm] = useState({
    cpf: "",
    password: "",
  });
  const { setIsLogged } = useContext(context);

  const router = useRouter();

  const handleSubmit: FormEventHandler = (e) => {
    e.preventDefault();
    setIsLogged(true);
    router.push("/dashboard");
  };

  const handleChange: ChangeEventHandler = (e) => {
    const { value, name } = e.target as HTMLInputElement;
    setForm((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  if (register) {
    return <Register setRegister={setRegister} />;
  }

  return (
    <Suspense fallback={<Spinner />}>
      <form
        onSubmit={handleSubmit}
        className="mt-10 flex w-[400px] flex-col justify-between gap-10 border px-3 py-4 max-sm:mx-3 max-sm:w-full"
      >
        <h2 className="text-lg font-bold">Login</h2>
        <label htmlFor="cpf" className="flex flex-col gap-2 text-sm">
          CPF
          <input
            type="text"
            name="cpf"
            id="cpf"
            value={form.cpf}
            onChange={handleChange}
            className="rounded-md border-none bg-gray-50"
          />
        </label>
        <label
          htmlFor="password"
          className="relative flex flex-col gap-2 text-sm"
        >
          Senha
          <input
            type={show ? "text" : "password"}
            name="password"
            id="password"
            value={form.password}
            onChange={handleChange}
            className="rounded-md border-none bg-gray-50"
          />
          <button
            type="button"
            onClick={() => setShow(!show)}
            className="absolute bottom-3 right-2"
          >
            {show ? <BsEyeSlash /> : <BsEye />}
          </button>
        </label>
        <button
          type="submit"
          className="rounded-md bg-green-600 px-6 py-2.5 font-sans text-sm font-semibold text-white"
        >
          ENTRAR
        </button>
        <p className="text-center font-sans text-xs">
          Ainda n√£o sou cadastrado.{" "}
          <a
            className="cursor-pointer text-blue-100 underline"
            onClick={() => setRegister(true)}
          >
            Fazer cadastro
          </a>
        </p>
      </form>
    </Suspense>
  );
};

export default Login;
