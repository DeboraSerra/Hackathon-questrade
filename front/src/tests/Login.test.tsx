import Login from "../components/Login/Login"
import { render } from '@testing-library/react'

jest.mock('next/navigation', () => ({
  useRouter: jest.fn().mockReturnValue({
    query: {},
    push: () => {}
  })
}))

jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    return <img {...props} />
  },
}))

describe('Test the login page', () => {
  it('Should match the snapshot', () => {
    const { container } = render(<Login />)
    expect(container).toMatchSnapshot()
  })
})