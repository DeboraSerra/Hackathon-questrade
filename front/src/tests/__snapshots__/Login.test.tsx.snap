// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the login page Should match the snapshot 1`] = `
<div>
  <form
    class="mt-10 flex w-[400px] flex-col justify-between gap-10 border px-3 py-4 max-sm:mx-3 max-sm:w-full"
  >
    <h2
      class="text-lg font-bold"
    >
      Login
    </h2>
    <label
      class="flex flex-col gap-2 text-sm"
      for="email"
    >
      E-mail
      <input
        class="rounded-md border-none bg-gray-50"
        id="email"
        name="email"
        type="email"
        value=""
      />
    </label>
    <label
      class="relative flex flex-col gap-2 text-sm"
      for="password"
    >
      Senha
      <input
        class="rounded-md border-none bg-gray-50"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        class="absolute bottom-3 right-2"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 8s-3-5.5-8-5.5S0 8 0 8s3 5.5 8 5.5S16 8 16 8zM1.173 8a13.133 13.133 0 0 1 1.66-2.043C4.12 4.668 5.88 3.5 8 3.5c2.12 0 3.879 1.168 5.168 2.457A13.133 13.133 0 0 1 14.828 8c-.058.087-.122.183-.195.288-.335.48-.83 1.12-1.465 1.755C11.879 11.332 10.119 12.5 8 12.5c-2.12 0-3.879-1.168-5.168-2.457A13.134 13.134 0 0 1 1.172 8z"
          />
          <path
            d="M8 5.5a2.5 2.5 0 1 0 0 5 2.5 2.5 0 0 0 0-5zM4.5 8a3.5 3.5 0 1 1 7 0 3.5 3.5 0 0 1-7 0z"
          />
        </svg>
      </button>
    </label>
    <button
      class="rounded-md bg-green-600 px-6 py-2.5 font-sans text-sm font-semibold text-white"
      type="submit"
    >
      ENTRAR
    </button>
    <p
      class="text-center font-sans text-xs"
    >
      Ainda n√£o sou cadastrado.
       
      <a
        class="cursor-pointer text-blue-100 underline"
      >
        Fazer cadastro
      </a>
    </p>
  </form>
</div>
`;
